go/vt/vtgate/vstream_manager.go

```diff
@@ -404,25 +418,66 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 
 		var eventss [][]*binlogdatapb.VEvent
 		var err error
-		rss := vs.rss
-		if vs.resolver != nil {
-			rss, err = vs.resolver.ResolveDestination(ctx, sgtid.Keyspace, vs.tabletType, key.DestinationShard(sgtid.Shard))
-			if err != nil {
-				return err
-			}
+		tp, err := discovery.NewTabletPicker(vs.ts, []string{vs.vsm.cell}, sgtid.Keyspace, sgtid.Shard, vs.tabletType.String())
+		if err != nil {
+			log.Errorf(err.Error())
+			return err
 		}
-		if len(rss) != 1 {
-			// Unreachable.
-			return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "unexpected number or shards: %v", rss)
+		tablet, err := tp.PickForStreaming(ctx)
+		if err != nil {
+			log.Errorf(err.Error())
+			return err
 		}
+		log.Infof("Picked tablet %s for for %s/%s/%s/%s", tablet.Alias.String(), vs.vsm.cell, sgtid.Keyspace, sgtid.Shard, vs.tabletType.String())
+		target := &querypb.Target{
+			Keyspace:   sgtid.Keyspace,
+			Shard:      sgtid.Shard,
+			TabletType: vs.tabletType,
+			Cell:       vs.vsm.cell,
+		}
+		tabletConn, err := vs.vsm.resolver.GetGateway().QueryServiceByAlias(tablet.Alias, target)
+		if err != nil {
+			log.Errorf(err.Error())
+			return err
+		}
+
+		errCh := make(chan string, 1)
+		go func() {
+			tabletConn.StreamHealth(ctx, func(shr *querypb.StreamHealthResponse) error {
+				var errString string
+				if ctx.Err() != nil {
+					return fmt.Errorf("context has ended")
+				}
+				if shr == nil || shr.RealtimeStats == nil || shr.Target == nil {
+					return fmt.Errorf("health check failed")
+				}
+				if vs.tabletType != shr.Target.TabletType {
+					errString = fmt.Sprintf("tablet %s is no longer healthy: %s, restarting vstream",
```

Comment:
Shouldn't the error string here be the one below?
```
errString = fmt.Sprintf("tablet type has changed from %s to %s, restarting vstream",
```

Reply:
ouch, will fix this